### PR TITLE
Issue #2401: GNIP: GeoNode Backup and Restore

### PR DESCRIPTION
**Pull request by afabiani**
GeoNode Backup and Restore Management Commands

### CODE REVIEW

File: geonode/base/management/commands/backup.py [hunk 1]
```diff
@@ -0,0 +1,200 @@
+#########################################################################
+#
+# Copyright (C) 2016 OpenPlans
```
Comment (jj0hns0n) [inline geonode/base/management/commands/backup.py]: Lets change these to OSGeo now

File: geonode/base/management/commands/helpers.py [hunk 2]
```diff
@@ -0,0 +1,204 @@
+from __future__ import with_statement
```
Comment (jj0hns0n) [inline geonode/base/management/commands/helpers.py]: Please add a header

File: geonode/base/management/commands/lib/gn20_to_24.py [hunk 3]
```diff
@@ -0,0 +1,261 @@
+import re
```
Comment (jj0hns0n) [inline geonode/base/management/commands/lib/gn20_to_24.py]: please add a header

File: geonode/utils.py [hunk 4]
```diff
@@ -681,7 +681,8 @@ def check_shp_columnnames(layer):
                         j += 1
                         new_field_name = new_field_name[:-2]+'_'+str(j)
                 list_col.update({field_name: new_field_name})
-    except UnicodeDecodeError:
+    except UnicodeDecodeError as e:
+        print str(e)
```
Comment (jj0hns0n) [inline geonode/utils.py]: should probably us logging here

Comment (ingenieroariel) [inline geonode/utils.py]: Let's use the logging framework instead of printing to stdout.

File: scripts/backup-restore/migrate_layers.py [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+               obj = helpers.load_fixture(app_name, fixture_file, mangler=mangler, basepk=higher_pk, owner=owner, datastore=settings.OGC_SERVER['default']['DATASTORE'], siteurl=settings.SITEURL)
+
+               from django.core import serializers
+
+               objects = serializers.deserialize('json', json.dumps(obj), ignorenonexistent=True)
+               for obj in objects:
+                  obj.save(using=DEFAULT_DB_ALIAS)
+
+   except Exception, err:
+      #print str(err)
```
Comment (jj0hns0n) [inline scripts/backup-restore/migrate_layers.py]: lets use logging framework here?

**Comment by ingenieroariel** [comment]
Thanks @afabiani for this!.

My suggestion is to have one person try it at least from 2.4 final to another fresh 2.4 final and from 2.4 final to 2.5 latest and share with us a screen capture of how it works before merging it. @waybarrios could you help with that?

**Comment by waybarrios** [comment]
@ingenieroariel  I can help you with that! If I have troubles or comments, I'll let you know as soon as possible

**Comment by afabiani** [comment]
If you need we setup a public test machine at 188.165.62.190

If you need access please contact me privately.

Best Regards,
Alessio Fabiani.

==
GeoServer Professional Services from the experts!
Visit http://goo.gl/it488V for more information.
==

Ing. Alessio Fabiani
@alfa7691
Founder/Technical Lead

GeoSolutions S.A.S.
Via di Montramito 3/A
55054  Massarosa (LU)
Italy
phone: +39 0584 962313
fax:     +39 0584 1660272
mob:   +39 331 6233686

http://www.geo-solutions.it
http://twitter.com/geosolutions_it

-------------------------------------------------------

*AVVERTENZE AI SENSI DEL D.Lgs. 196/2003*

Le informazioni contenute in questo messaggio di posta elettronica e/o
nel/i file/s allegato/i sono da considerarsi strettamente riservate. Il
loro utilizzo è consentito esclusivamente al destinatario del messaggio,
per le finalità indicate nel messaggio stesso. Qualora riceviate questo
messaggio senza esserne il destinatario, Vi preghiamo cortesemente di
darcene notizia via e-mail e di procedere alla distruzione del messaggio
stesso, cancellandolo dal Vostro sistema. Conservare il messaggio stesso,
divulgarlo anche in parte, distribuirlo ad altri soggetti, copiarlo, od
utilizzarlo per finalità diverse, costituisce comportamento contrario ai
principi dettati dal D.Lgs. 196/2003.



The information in this message and/or attachments, is intended solely for
the attention and use of the named addressee(s) and may be confidential or
proprietary in nature or covered by the provisions of privacy act
(Legislative Decree June, 30 2003, no.196 - Italy's New Data Protection
Code).Any use not in accord with its purpose, any disclosure, reproduction,
copying, distribution, or either dissemination, either whole or partial, is
strictly forbidden except previous formal approval of the named
addressee(s). If you are not the intended recipient, please contact
immediately the sender by telephone, fax or e-mail and delete the
information in this message that has been received in error. The sender
does not give any warranty or accept liability as the content, accuracy or
completeness of sent messages and accepts no responsibility  for changes
made after they were sent or for other risks which arise as a result of
e-mail transmission, viruses, etc.

---------------------------------------------------------------------

**Comment by waybarrios** [comment]
@afabiani Of course! Thank you

**Comment by afabiani** [comment]
Hi Way, please ask for access at 188.165.62.190 to Alessandro Parma <
alessandro.parma@geo-solutions.it>

You will receive both ssh access and geonode credentials.

About the backup and restore commands, for now it is possible to execute
them only through the managemend command via shell:

1. Login into the machine
2. Go to -> cd /home/geonode/geonode/

Backup command:

python manage.py backup --backup-dir=/tmp/


at the end of the backup a new archive will be created automatically. The
command line will tell you which is the full path of the archive


Restore command:

*sudo *python manage.py restore --backup-file=/tmp/2016-09-14_032126.zip

*WARNING:*
When performing those commands

 1. The backend (DB or whatever) is accessible and you have rights
 2. The GeoServer is up and running and reachable from this machine

*WARNING:* The restore will overwrite ALL GeoNode data and GeoServer
configuration.



Best Regards,
Alessio Fabiani.

==
GeoServer Professional Services from the experts!
Visit http://goo.gl/it488V for more information.
==

Ing. Alessio Fabiani
@alfa7691
Founder/Technical Lead

GeoSolutions S.A.S.
Via di Montramito 3/A
55054  Massarosa (LU)
Italy
phone: +39 0584 962313
fax:     +39 0584 1660272
mob:   +39 331 6233686

http://www.geo-solutions.it
http://twitter.com/geosolutions_it

-------------------------------------------------------

*AVVERTENZE AI SENSI DEL D.Lgs. 196/2003*

Le informazioni contenute in questo messaggio di posta elettronica e/o
nel/i file/s allegato/i sono da considerarsi strettamente riservate. Il
loro utilizzo è consentito esclusivamente al destinatario del messaggio,
per le finalità indicate nel messaggio stesso. Qualora riceviate questo
messaggio senza esserne il destinatario, Vi preghiamo cortesemente di
darcene notizia via e-mail e di procedere alla distruzione del messaggio
stesso, cancellandolo dal Vostro sistema. Conservare il messaggio stesso,
divulgarlo anche in parte, distribuirlo ad altri soggetti, copiarlo, od
utilizzarlo per finalità diverse, costituisce comportamento contrario ai
principi dettati dal D.Lgs. 196/2003.



The information in this message and/or attachments, is intended solely for
the attention and use of the named addressee(s) and may be confidential or
proprietary in nature or covered by the provisions of privacy act
(Legislative Decree June, 30 2003, no.196 - Italy's New Data Protection
Code).Any use not in accord with its purpose, any disclosure, reproduction,
copying, distribution, or either dissemination, either whole or partial, is
strictly forbidden except previous formal approval of the named
addressee(s). If you are not the intended recipient, please contact
immediately the sender by telephone, fax or e-mail and delete the
information in this message that has been received in error. The sender
does not give any warranty or accept liability as the content, accuracy or
completeness of sent messages and accepts no responsibility  for changes
made after they were sent or for other risks which arise as a result of
e-mail transmission, viruses, etc.

---------------------------------------------------------------------

**Comment by afabiani** [comment]
A small add to this.

On a new setup the step 0 is to take a look at this very easy and straight
forward configuration file

geonode/base/management/commands/settings.ini

full path on the test machine

/home/geonode/geonode/geonode/base/management/commands/settings.ini



Best Regards,
Alessio Fabiani.

==
GeoServer Professional Services from the experts!
Visit http://goo.gl/it488V for more information.
==

Ing. Alessio Fabiani
@alfa7691
Founder/Technical Lead

GeoSolutions S.A.S.
Via di Montramito 3/A
55054  Massarosa (LU)
Italy
phone: +39 0584 962313
fax:     +39 0584 1660272
mob:   +39 331 6233686

http://www.geo-solutions.it
http://twitter.com/geosolutions_it

-------------------------------------------------------

*AVVERTENZE AI SENSI DEL D.Lgs. 196/2003*

Le informazioni contenute in questo messaggio di posta elettronica e/o
nel/i file/s allegato/i sono da considerarsi strettamente riservate. Il
loro utilizzo è consentito esclusivamente al destinatario del messaggio,
per le finalità indicate nel messaggio stesso. Qualora riceviate questo
messaggio senza esserne il destinatario, Vi preghiamo cortesemente di
darcene notizia via e-mail e di procedere alla distruzione del messaggio
stesso, cancellandolo dal Vostro sistema. Conservare il messaggio stesso,
divulgarlo anche in parte, distribuirlo ad altri soggetti, copiarlo, od
utilizzarlo per finalità diverse, costituisce comportamento contrario ai
principi dettati dal D.Lgs. 196/2003.



The information in this message and/or attachments, is intended solely for
the attention and use of the named addressee(s) and may be confidential or
proprietary in nature or covered by the provisions of privacy act
(Legislative Decree June, 30 2003, no.196 - Italy's New Data Protection
Code).Any use not in accord with its purpose, any disclosure, reproduction,
copying, distribution, or either dissemination, either whole or partial, is
strictly forbidden except previous formal approval of the named
addressee(s). If you are not the intended recipient, please contact
immediately the sender by telephone, fax or e-mail and delete the
information in this message that has been received in error. The sender
does not give any warranty or accept liability as the content, accuracy or
completeness of sent messages and accepts no responsibility  for changes
made after they were sent or for other risks which arise as a result of
e-mail transmission, viruses, etc.

---------------------------------------------------------------------

**Comment by waybarrios** [comment]
@ingenieroariel Until now, I tried to backup a geonode with layers using as reference geo-solutions server. 

**Backup process:**
test1

When I did the backup, a zip file was created which contains json files with layers, maps, docs, and accounts info. It also has static folder, template folder and geoserver data. You can see the following screenshot:
test3

After that, it merged all data to new fresh geonode!.

**Comment by jj0hns0n** [comment]
Can you post some stacktrace of what the restore looks like?

**Comment by waybarrios** [comment]
@jj0hns0n  I will do! 

What I can show now, it is the restore process by screen capture. It restored San Andres and Providence Layer (same layer that I used before). I also loaded previous information.

resto1

**Comment by afabiani** [comment]
image

File: geonode/base/management/commands/helpers.py (hunk 2 above)
Comment (afabiani) [inline geonode/base/management/commands/helpers.py]: Fixed!

File: geonode/base/management/commands/lib/gn20_to_24.py (hunk 3 above)
Comment (afabiani) [inline geonode/base/management/commands/lib/gn20_to_24.py]: Fixed!

File: geonode/utils.py (hunk 4 above)
Comment (afabiani) [inline geonode/utils.py]: fixed

File: scripts/backup-restore/migrate_layers.py (hunk 5 above)
Comment (afabiani) [inline scripts/backup-restore/migrate_layers.py]: fixed

**Comment by afabiani** [comment]
Added Documentation to Admin docs for Backup & Restore.

GeoServer full documentation is also available here

http://docs.geoserver.org/latest/en/user/community/backuprestore/index.html

File: geonode/base/management/commands/backup.py [hunk 6]
```diff
@@ -0,0 +1,240 @@
...
+        print " 2. The GeoServer is up and running and reachable from this machine"
+        message = 'You want to proceed?'
+
+        if force_exec or helpers.confirm(prompt=message, resp=False):
+            # Create Target Folder
+            dir_time_suffix = helpers.get_dir_time_suffix()
+            target_folder = os.path.join(backup_dir, dir_time_suffix)
+            if not os.path.exists(target_folder):
+                os.makedirs(target_folder)
+            os.chmod(target_folder, 0777)
```
Comment (kalxas) [inline geonode/base/management/commands/backup.py]: 777 is always a bad idea

File: geonode/base/models.py [hunk 7]
```diff
@@ -64,7 +64,7 @@ class ContactRole(models.Model):
     """
     ContactRole is an intermediate model to bind Profiles as Contacts to Resources and apply roles.
     """
-    resource = models.ForeignKey('ResourceBase')
+    resource = models.ForeignKey('ResourceBase', blank=True, null=True)
```
Comment (kalxas) [inline geonode/base/models.py]: Is this change here to avoid errors during restore?

File: geonode/base/models.py [hunk 8]
```diff
@@ -804,7 +822,7 @@ class Link(models.Model):
         * OGC:WFS: for WFS service links
         * OGC:WCS: for WCS service links
     """
-    resource = models.ForeignKey(ResourceBase)
+    resource = models.ForeignKey(ResourceBase, blank=True, null=True)
```
Comment (kalxas) [inline geonode/base/models.py]: Same as above

File: geonode/base/management/commands/backup.py (hunk 6 above)
Comment (afabiani) [inline geonode/base/management/commands/backup.py]: Good point, I'll change it. Thanks for suggesting.

File: geonode/base/models.py (hunk 7 above)
Comment (afabiani) [inline geonode/base/models.py]: Yeps. Currently GeoNode has a hierarchical model and there is no way to allow it recognize that a Layer is "also" a resourcebase by using fixtures. Unfortunately this is a core issue of DJango and I honestly did not found any solution except relaxing the model.

File: geonode/base/models.py (hunk 8 above)
Comment (afabiani) [inline geonode/base/models.py]: Same as above.

It is worth notice that this change does not impact GeoNode at all.

**Comment by afabiani** [comment]
Hi all, @jj0hns0n how about the review? Can we move forward with this?

**Pull request merged**